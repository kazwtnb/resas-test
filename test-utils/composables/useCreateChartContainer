export const useCreateChartContainer = () => {
  const container = document.createElement('div');
  const chartTarget = document.createElement('div');
  chartTarget.id = 'test-chart-container';
  container.appendChild(chartTarget);

  return { container, chartTarget };
};
