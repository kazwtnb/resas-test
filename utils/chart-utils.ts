import { Options } from 'highcharts';
import Highcharts from 'highcharts/highcharts.js';
import { CreateLineChartInput } from '../types/models/chart-model';

/**
 * Highchartsで線グラフを描画する
 */
export const createLineChart = (input: CreateLineChartInput): void => {
  Highcharts.chart(input.renderTo, {
    title: {
      text: ''
    },
    xAxis: {
      title: {
        text: input.xAxisTitle,
        align: 'high',
        rotation: 0,
        x: 25,
        y: -18
      }
    },
    yAxis: {
      title: {
        text: input.yAxisTitle,
        align: 'high',
        rotation: 0,
        y: -20,
        offset: -20
      }
    },
    chart: {
      marginTop: 40,
      marginRight: 20
    },
    plotOptions: {
      series: {
        label: {
          connectorAllowed: false
        },
        pointStart: input.startYear,
        pointInterval: input.yearInterval
      }
    },
    series: input.data,
    responsive: {
      rules: [
        {
          condition: {
            maxWidth: 500
          },
          chartOptions: {
            legend: {
              layout: 'horizontal',
              align: 'center',
              verticalAlign: 'bottom'
            }
          }
        }
      ]
    }
  } as Options);
};

/**
 * 固定値のIDを使用する設定に変更する
 * snapshotテストが自動採番IDで落ちるので追加
 */
export const useTestConfig = () => {
  Highcharts.useSerialIds(true);
};
