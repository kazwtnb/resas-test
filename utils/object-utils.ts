import { isEqual, pick } from 'lodash-es';

export const isEqualObjectByKey = <T>(
  obj1: T,
  obj2: T,
  idKey?: string
): boolean =>
    idKey
      ? isEqual(pick(obj1, [idKey]), pick(obj2, [idKey]))
      : isEqual(obj1, obj2);
