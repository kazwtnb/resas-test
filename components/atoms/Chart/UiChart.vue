<template>
  <div class="ui-chart">
    <div id="ui-chart-container" />
    <span
      v-if="!!emptyMessage && !data.length"
      class="ui-chart-empty-message"
      >{{ emptyMessage }}</span
    >
  </div>
</template>
<script setup lang="ts">
import { toRefs, onMounted, watch } from 'vue';
import { ChartData } from '~~/types/models/chart-model';
import { createLineChart } from '~~/utils/chart-utils';

interface Props {
  renderTo?: HTMLDivElement;
  xAxisTitle?: string;
  yAxisTitle?: string;
  startYear?: number;
  yearInterval?: number;
  emptyMessage?: string;
  data: ChartData[];
}

const props = withDefaults(defineProps<Props>(), {
  renderTo: undefined,
  xAxisTitle: '',
  yAxisTitle: '',
  startYear: 2000,
  yearInterval: 5,
  emptyMessage: ''
});
const { data } = toRefs(props);

const createChart = () => {
  createLineChart({
    renderTo: props.renderTo || 'ui-chart-container',
    xAxisTitle: props.xAxisTitle,
    yAxisTitle: props.yAxisTitle,
    startYear: props.startYear,
    yearInterval: props.yearInterval,
    data: data.value
  });
};

onMounted(() => {
  createChart();
});

watch(data, () => {
  createChart();
});
</script>
<style lang="scss" scoped>
.ui-chart {
  position: relative;
  color: colors.$color-border;
  .ui-chart-empty-message {
    position: absolute;
    top: 50%;
    left: 50%;
    transform: translate(-50%, -50%);
  }
}
</style>
