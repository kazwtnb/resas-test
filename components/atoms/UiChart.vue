<template>
  <div :id="props.id" class="ui-chart" />
</template>
<script setup lang="ts">
import { ChartData } from '~~/types/models/chart-model';
import { createLineChart } from '~~/utils/chart-utils';

interface Props {
  id?: string;
  xAxisTitle?: string;
  yAxisTitle?: string;
  startYear?: number;
  yearInterval?: number;
  data: ChartData[];
}

const props = withDefaults(defineProps<Props>(), {
  id: 'ui-chart-container',
  xAxisTitle: '',
  yAxisTitle: '',
  startYear: 2001,
  yearInterval: 5
});
const { data } = toRefs(props);

const createChart = () => {
  createLineChart({
    id: props.id,
    xAxisTitle: props.xAxisTitle,
    yAxisTitle: props.yAxisTitle,
    startYear: props.startYear,
    yearInterval: props.yearInterval,
    data: data.value
  });
};

onMounted(() => {
  createChart();
});

watch(data, () => {
  createChart();
});
</script>
<style lang="scss" scoped></style>
