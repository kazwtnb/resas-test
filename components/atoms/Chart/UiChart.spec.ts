import { mount } from '@vue/test-utils';
import { describe, expect, test, beforeAll } from 'vitest';
import UiChart from './UiChart.vue';
import { useCreateChartContainer } from '~~/test-utils/composables/useCreateChartContainer';
import { useTestConfig } from '~~/utils/chart-utils';

describe('UiChart', () => {
  beforeAll(() => {
    useTestConfig();
  });

  describe('snapshot', () => {
    test.each([
      {
        name: '複数データあり',
        input: [
          { name: 'a', data: [1, 2, 3, 4, 5] },
          { name: 'b', data: [3, 4, 5, 6, 7] }
        ]
      },
      {
        name: 'empty表示',
        input: []
      }
    ])(`%o`, ({ input }) => {
      const { container, chartTarget } = useCreateChartContainer();

      const wrapper = mount(UiChart, {
        attachTo: container,
        props: {
          renderTo: chartTarget,
          xAxisTitle: 'x軸',
          yAxisTitle: 'y軸',
          startYear: 2000,
          yearInterval: 5,
          emptyMessage: 'empty-state',
          data: input
        }
      });
      expect(container).toMatchSnapshot();
    });
  });
});
