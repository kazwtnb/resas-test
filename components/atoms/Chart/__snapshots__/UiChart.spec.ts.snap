// Vitest Snapshot v1

exports[`UiChart > snapshot > {
  name: '複数データあり',
  input: [
    { name: 'a', data: [ 1, 2, 3, 4, 5, [length]: 5 ] },
    { name: 'b', data: [ 3, 4, 5, 6, 7, [length]: 5 ] },
    [length]: 2
  ]
} 1`] = `
<div>
  <div
    data-highcharts-chart="0"
    id="test-chart-container"
    style="overflow: hidden;"
  >
    <div
      class="highcharts-container "
      dir="ltr"
      id="highcharts-0"
      style="position: relative; overflow: hidden; width: 600px; height: 400px; text-align: left; line-height: normal; z-index: 0; user-select: none; outline: none;"
    >
      <svg
        class="highcharts-root"
        height="400"
        style="font-family: \\"Lucida Grande\\", \\"Lucida Sans Unicode\\", Arial, Helvetica, sans-serif; font-size: 12px;"
        version="1.1"
        viewBox="0 0 600 400"
        width="600"
        xmlns="http://www.w3.org/2000/svg"
      >
        <desc>
          Created with Highcharts 10.1.0
        </desc>
        <defs>
          <clippath
            id="highcharts-1-"
          >
            <rect
              fill="none"
              height="NaN"
              width="NaN"
              x="0"
              y="0"
            />
          </clippath>
          <clippath
            id="highcharts-7-"
          >
            <rect
              fill="none"
              height="NaNNaN"
              width="NaNNaN"
              x="0"
              y="0"
            />
          </clippath>
          <clippath
            id="highcharts-8-"
          >
            <rect
              class="highcharts-animating"
              fill="none"
              height="NaN"
              width="NaN"
              x="0"
              y="0"
            />
          </clippath>
          <clippath
            id="highcharts-9-"
          >
            <rect
              fill="none"
              height="400"
              width="99"
              x="-99"
              y="-40"
            />
          </clippath>
        </defs>
        <rect
          class="highcharts-background"
          fill="#ffffff"
          height="400"
          rx="0"
          ry="0"
          width="600"
          x="0"
          y="0"
        />
        <rect
          class="highcharts-plot-background"
          fill="none"
          height="NaN"
          width="NaN"
          x="NaN"
          y="40"
        />
        <g
          class="highcharts-grid highcharts-xaxis-grid"
          data-z-index="1"
        >
          <path
            class="highcharts-grid-line"
            d="M 0 0"
            data-z-index="1"
            fill="none"
            opacity="1"
            stroke-dasharray="none"
          />
          <path
            class="highcharts-grid-line"
            d="M 0 0"
            data-z-index="1"
            fill="none"
            opacity="1"
            stroke-dasharray="none"
          />
          <path
            class="highcharts-grid-line"
            d="M 0 0"
            data-z-index="1"
            fill="none"
            opacity="1"
            stroke-dasharray="none"
          />
          <path
            class="highcharts-grid-line"
            d="M 0 0"
            data-z-index="1"
            fill="none"
            opacity="1"
            stroke-dasharray="none"
          />
          <path
            class="highcharts-grid-line"
            d="M 0 0"
            data-z-index="1"
            fill="none"
            opacity="1"
            stroke-dasharray="none"
          />
        </g>
        <g
          class="highcharts-grid highcharts-yaxis-grid"
          data-z-index="1"
        >
          <path
            class="highcharts-grid-line"
            d="M 0 0"
            data-z-index="1"
            fill="none"
            opacity="1"
            stroke="#e6e6e6"
            stroke-dasharray="none"
            stroke-width="1"
          />
          <path
            class="highcharts-grid-line"
            d="M 0 0"
            data-z-index="1"
            fill="none"
            opacity="1"
            stroke="#e6e6e6"
            stroke-dasharray="none"
            stroke-width="1"
          />
          <path
            class="highcharts-grid-line"
            d="M 0 0"
            data-z-index="1"
            fill="none"
            opacity="1"
            stroke="#e6e6e6"
            stroke-dasharray="none"
            stroke-width="1"
          />
          <path
            class="highcharts-grid-line"
            d="M 0 0"
            data-z-index="1"
            fill="none"
            opacity="1"
            stroke="#e6e6e6"
            stroke-dasharray="none"
            stroke-width="1"
          />
          <path
            class="highcharts-grid-line"
            d="M 0 0"
            data-z-index="1"
            fill="none"
            opacity="1"
            stroke="#e6e6e6"
            stroke-dasharray="none"
            stroke-width="1"
          />
        </g>
        <rect
          class="highcharts-plot-border"
          data-z-index="1"
          fill="none"
          height="0"
          width="0"
          x="0"
          y="40"
        />
        <g
          class="highcharts-axis highcharts-xaxis"
          data-z-index="2"
        >
          <path
            class="highcharts-tick"
            d="M 0 0"
            fill="none"
            opacity="1"
            stroke="#ccd6eb"
            stroke-width="1"
          />
          <path
            class="highcharts-tick"
            d="M 0 0"
            fill="none"
            opacity="1"
            stroke="#ccd6eb"
            stroke-width="1"
          />
          <path
            class="highcharts-tick"
            d="M 0 0"
            fill="none"
            opacity="1"
            stroke="#ccd6eb"
            stroke-width="1"
          />
          <path
            class="highcharts-tick"
            d="M 0 0"
            fill="none"
            opacity="1"
            stroke="#ccd6eb"
            stroke-width="1"
          />
          <path
            class="highcharts-tick"
            d="M 0 0"
            fill="none"
            opacity="1"
            stroke="#ccd6eb"
            stroke-width="1"
          />
          <text
            class="highcharts-axis-title"
            data-z-index="7"
            style="color: rgb(102, 102, 102); fill: #666666;"
            text-anchor="end"
            transform="translate(0,0)"
            x="NaN"
            y="NaN"
          >
            x軸
          </text>
          <path
            class="highcharts-axis-line"
            d="M 0 0"
            data-z-index="7"
            fill="none"
            stroke="#ccd6eb"
            stroke-width="1"
          />
        </g>
        <g
          class="highcharts-axis highcharts-yaxis"
          data-z-index="2"
        >
          <text
            class="highcharts-axis-title"
            data-z-index="7"
            style="color: rgb(102, 102, 102); fill: #666666;"
            text-anchor="end"
            transform="translate(0,0)"
            x="NaN"
            y="20"
          >
            y軸
          </text>
          <path
            class="highcharts-axis-line"
            d="M 0 0"
            data-z-index="7"
            fill="none"
          />
        </g>
        <g
          class="highcharts-series-group"
          data-z-index="3"
        >
          <g
            class="highcharts-series highcharts-series-0 highcharts-line-series highcharts-color-0"
            clip-path="url(#highcharts-8-)"
            data-z-index="0.1"
            opacity="1"
            transform="translate(0,40) scale(1 1)"
          >
            <path
              class="highcharts-graph"
              d="M -100000 -100000 L -100000 -100000 L -100000 -100000 L -100000 -100000 L -100000 -100000"
              data-z-index="1"
              fill="none"
              stroke="#7cb5ec"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
            />
            <path
              class="highcharts-tracker-line"
              d="M -100000 -100000 L -100000 -100000 L -100000 -100000 L -100000 -100000 L -100000 -100000"
              data-z-index="2"
              fill="none"
              stroke="rgba(192,192,192,0.002)"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="22"
            />
          </g>
          <g
            class="highcharts-markers highcharts-series-0 highcharts-line-series highcharts-color-0 highcharts-tracker"
            clip-path="url(#highcharts-9-)"
            data-z-index="0.1"
            opacity="1"
            transform="translate(0,40) scale(1 1)"
          />
          <g
            class="highcharts-series highcharts-series-1 highcharts-line-series highcharts-color-1"
            clip-path="url(#highcharts-8-)"
            data-z-index="0.1"
            opacity="1"
            transform="translate(0,40) scale(1 1)"
          >
            <path
              class="highcharts-graph"
              d="M -100000 -100000 L -100000 -100000 L -100000 -100000 L -100000 -100000 L -100000 -100000"
              data-z-index="1"
              fill="none"
              stroke="#434348"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
            />
            <path
              class="highcharts-tracker-line"
              d="M -100000 -100000 L -100000 -100000 L -100000 -100000 L -100000 -100000 L -100000 -100000"
              data-z-index="2"
              fill="none"
              stroke="rgba(192,192,192,0.002)"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="22"
            />
          </g>
          <g
            class="highcharts-markers highcharts-series-1 highcharts-line-series highcharts-color-1 highcharts-tracker"
            clip-path="url(#highcharts-9-)"
            data-z-index="0.1"
            opacity="1"
            transform="translate(0,40) scale(1 1)"
          />
        </g>
        <text
          class="highcharts-title"
          data-z-index="4"
          style="color: rgb(51, 51, 51); font-size: 18px; fill: #333333;"
          text-anchor="middle"
          x="300"
          y="24"
        />
        <text
          class="highcharts-subtitle"
          data-z-index="4"
          style="color: rgb(102, 102, 102); fill: #666666;"
          text-anchor="middle"
          x="300"
          y="10"
        />
        <text
          class="highcharts-caption"
          data-z-index="4"
          style="color: rgb(102, 102, 102); fill: #666666;"
          text-anchor="start"
          x="10"
          y="397"
        />
        <g
          class="highcharts-legend highcharts-no-tooltip"
          data-z-index="7"
          transform="translate(300,359)"
        >
          <rect
            class="highcharts-legend-box"
            fill="none"
            rx="0"
            ry="0"
          />
          <g
            data-z-index="1"
          >
            <g>
              <g
                class="highcharts-legend-item highcharts-line-series highcharts-color-0 highcharts-series-0"
                data-z-index="1"
                transform="translate(8,3)"
              >
                <path
                  class="highcharts-graph"
                  d="M 0 11 L 16 11"
                  fill="none"
                  stroke="#7cb5ec"
                  stroke-width="2"
                />
                <path
                  class="highcharts-point"
                  d="M 8 15 A 4 4 0 1 1 8.003999999333336 14.999998000000167 Z"
                  fill="#7cb5ec"
                  opacity="1"
                />
                <text
                  data-z-index="2"
                  style="color: rgb(51, 51, 51); cursor: pointer; font-size: 12px; font-weight: bold; fill: #333333;"
                  text-anchor="start"
                  x="21"
                  y="15"
                >
                  a
                </text>
              </g>
              <g
                class="highcharts-legend-item highcharts-line-series highcharts-color-1 highcharts-series-1"
                data-z-index="1"
                transform="translate(0,3)"
              >
                <path
                  class="highcharts-graph"
                  d="M 0 11 L 16 11"
                  fill="none"
                  stroke="#434348"
                  stroke-width="2"
                />
                <path
                  class="highcharts-point"
                  d="M 8 7 L 12 11 L 8 15 L 4 11 Z"
                  fill="#434348"
                  opacity="1"
                />
                <text
                  data-z-index="2"
                  style="color: rgb(51, 51, 51); cursor: pointer; font-size: 12px; font-weight: bold; fill: #333333;"
                  text-anchor="start"
                  x="21"
                  y="15"
                >
                  b
                </text>
              </g>
            </g>
          </g>
        </g>
        <g
          class="highcharts-axis-labels highcharts-xaxis-labels"
          data-z-index="7"
        >
          <text
            style="color: rgb(102, 102, 102); cursor: default; font-size: 11px; fill: #666666;"
            text-anchor="middle"
            transform="translate(0,0)"
            x="0"
          >
            2000
          </text>
          <text
            style="color: rgb(102, 102, 102); cursor: default; font-size: 11px; fill: #666666;"
            text-anchor="middle"
            transform="translate(0,0)"
            x="0"
          >
            2005
          </text>
          <text
            style="color: rgb(102, 102, 102); cursor: default; font-size: 11px; fill: #666666;"
            text-anchor="middle"
            transform="translate(0,0)"
            x="0"
          >
            2010
          </text>
          <text
            style="color: rgb(102, 102, 102); cursor: default; font-size: 11px; fill: #666666;"
            text-anchor="middle"
            transform="translate(0,0)"
            x="0"
          >
            2015
          </text>
          <text
            style="color: rgb(102, 102, 102); cursor: default; font-size: 11px; fill: #666666;"
            text-anchor="middle"
            transform="translate(0,0)"
            x="0"
          >
            2020
          </text>
        </g>
        <g
          class="highcharts-axis-labels highcharts-yaxis-labels"
          data-z-index="7"
        >
          <text
            style="color: rgb(102, 102, 102); cursor: default; font-size: 11px; fill: #666666;"
            text-anchor="end"
            transform="translate(0,0)"
            x="0"
          >
            0
          </text>
          <text
            style="color: rgb(102, 102, 102); cursor: default; font-size: 11px; fill: #666666;"
            text-anchor="end"
            transform="translate(0,0)"
            x="0"
          >
            2
          </text>
          <text
            style="color: rgb(102, 102, 102); cursor: default; font-size: 11px; fill: #666666;"
            text-anchor="end"
            transform="translate(0,0)"
            x="0"
          >
            4
          </text>
          <text
            style="color: rgb(102, 102, 102); cursor: default; font-size: 11px; fill: #666666;"
            text-anchor="end"
            transform="translate(0,0)"
            x="0"
          >
            6
          </text>
          <text
            style="color: rgb(102, 102, 102); cursor: default; font-size: 11px; fill: #666666;"
            text-anchor="end"
            transform="translate(0,0)"
            x="0"
          >
            8
          </text>
        </g>
        <text
          class="highcharts-credits"
          data-z-index="8"
          style="cursor: pointer; color: rgb(153, 153, 153); font-size: 9px; fill: #999999;"
          text-anchor="end"
          x="590"
          y="395"
        >
          Highcharts.com
        </text>
      </svg>
    </div>
  </div>
  <div
    data-v-app=""
  >
    <div
      class="ui-chart"
      data-v-4ef31541=""
    >
      <div
        data-v-4ef31541=""
        id="ui-chart-container"
      />
      <!--v-if-->
    </div>
  </div>
</div>
`;

exports[`UiChart > snapshot > { name: 'empty表示', input: [ [length]: 0 ] } 1`] = `
<div>
  <div
    data-highcharts-chart="1"
    id="test-chart-container"
    style="overflow: hidden;"
  >
    <div
      class="highcharts-container "
      dir="ltr"
      id="highcharts-16"
      style="position: relative; overflow: hidden; width: 600px; height: 400px; text-align: left; line-height: normal; z-index: 0; user-select: none; outline: none;"
    >
      <svg
        class="highcharts-root"
        height="400"
        style="font-family: \\"Lucida Grande\\", \\"Lucida Sans Unicode\\", Arial, Helvetica, sans-serif; font-size: 12px;"
        version="1.1"
        viewBox="0 0 600 400"
        width="600"
        xmlns="http://www.w3.org/2000/svg"
      >
        <desc>
          Created with Highcharts 10.1.0
        </desc>
        <defs>
          <clippath
            id="highcharts-17-"
          >
            <rect
              fill="none"
              height="345"
              width="570"
              x="0"
              y="0"
            />
          </clippath>
        </defs>
        <rect
          class="highcharts-background"
          fill="#ffffff"
          height="400"
          rx="0"
          ry="0"
          width="600"
          x="0"
          y="0"
        />
        <rect
          class="highcharts-plot-background"
          fill="none"
          height="345"
          width="570"
          x="10"
          y="40"
        />
        <rect
          class="highcharts-plot-border"
          data-z-index="1"
          fill="none"
          height="345"
          width="570"
          x="10"
          y="40"
        />
        <g
          class="highcharts-series-group"
          data-z-index="3"
        />
        <text
          class="highcharts-title"
          data-z-index="4"
          style="color: rgb(51, 51, 51); font-size: 18px; fill: #333333;"
          text-anchor="middle"
          x="300"
          y="24"
        />
        <text
          class="highcharts-subtitle"
          data-z-index="4"
          style="color: rgb(102, 102, 102); fill: #666666;"
          text-anchor="middle"
          x="300"
          y="10"
        />
        <text
          class="highcharts-caption"
          data-z-index="4"
          style="color: rgb(102, 102, 102); fill: #666666;"
          text-anchor="start"
          x="10"
          y="397"
        />
        <g
          class="highcharts-legend highcharts-no-tooltip"
          data-z-index="7"
        >
          <rect
            class="highcharts-legend-box"
            fill="none"
            height="8"
            rx="0"
            ry="0"
            visibility="hidden"
            width="8"
            x="0"
            y="0"
          />
          <g
            data-z-index="1"
          >
            <g />
          </g>
        </g>
        <text
          class="highcharts-credits"
          data-z-index="8"
          style="cursor: pointer; color: rgb(153, 153, 153); font-size: 9px; fill: #999999;"
          text-anchor="end"
          x="590"
          y="395"
        >
          Highcharts.com
        </text>
      </svg>
    </div>
  </div>
  <div
    data-v-app=""
  >
    <div
      class="ui-chart"
      data-v-4ef31541=""
    >
      <div
        data-v-4ef31541=""
        id="ui-chart-container"
      />
      <span
        class="ui-chart-empty-message"
        data-v-4ef31541=""
      >
        empty-state
      </span>
    </div>
  </div>
</div>
`;
